### PR TITLE
237 fix assign detection when object is passed as character

### LINKED ISSUE #237: `get_code_dependency()` can not find objects when passed as character into the assign function
```r
code <- c("assign('adae', teal.data::rADAE)", "assign('adsl', teal.data::rADSL)")
tdata <- eval_code(teal_data(), code)
datanames(tdata) <- c("adsl", "adae")
get_code(tdata, datanames = "adsl")
```

brings

```r
character(0)
Warning message:
In get_code_dependency(object@code, datanames) :
  Object(s) not found in code: adsl
```

where it could/should bring 

```r
"assign(\"adsl\", teal.data::rADSL)"
```

### PR DESCRIPTION
**Pull request by m7pr**
Closes #237 

Detected by @chlebowa in https://github.com/insightsengineering/teal.data/pull/234#issuecomment-1867613019

Now

```r
code <- c("assign('adae', teal.data::rADAE)", "assign('adsl', teal.data::rADSL)")
tdata <- eval_code(teal_data(), code)
datanames(tdata) <- c("adsl", "adae")
get_code(tdata, datanames = "adsl")
```

brings

```r
[1] "assign(\"adsl\", teal.data::rADSL)"
```

### CODE REVIEW

**Comment by m7pr** [comment]
Below is the code I used to test various specification of `assign` usage. The desired output is to have a dependency output of a form of `c('d', '<-', 'assign)'`

```r
devtools::load_all(".")

code_2_cpd <- function(code) {
  code <- parse(text = code, keep.source = TRUE)
  pd <- utils::getParseData(code)
  calls_pd <- extract_calls(pd)
  calls_pd
}

code_2_extract <- function(code) {
  call_pd <- code_2_cpd(code)[[1]]
  assign_call <- call_pd$token == "SYMBOL_FUNCTION_CALL" & call_pd$text == "assign"
  # Check if parameters were named.
  if (any(call_pd$token == "SYMBOL_SUB")) {
    params <- call_pd[call_pd$token == "SYMBOL_SUB", "text"]
    if ('x' %in% params) {
      pos <- which(params == 'x')
    } else {
      pos <- length(params) + 1
    }
  } else {
    # Object is the first entry after 'assign'.
    pos <- 1
  }
  sym <- call_pd[which(assign_call) + pos, "text"]
  return(c(gsub("'", "", sym), "<-", "assign"))
  
}

code_2_cpd(code)

code1  <- "assign('d', iris)"
code2  <- "assign('d', 'iris')"
code3  <- "assign(d, 'iris')"
code4  <- "assign(d, iris)"
code5  <- "assign(value = iris, x = 'd')"
code6  <- "assign(value = 'iris', x = 'd')"
code7  <- "assign(value = 'iris', x = d)"
code8  <- "assign(value = iris, x = d)"
code9  <- "assign(value = iris, 'd')"
code10 <- "assign(value = 'iris', 'd')"
code11 <- "assign(value = 'iris', d)"
code12 <- "assign(value = iris, d)"

# 1. Check if there was $token="SYMBOL_FUNCTION_CALL" & $text = "assign"
# 2. Check if parameters were named $token="SYMBOL_SUB", "text"]
# 3. Check first thing after 'assign' unless 
# 3.a) x was in named parameters - then check it's position
# 3.b) x was not in named parameters - then check length(SYMBOL_SUB) + 1


code_2_extract(code1)
code_2_extract(code2)
code_2_extract(code3)
code_2_extract(code4)
code_2_extract(code5)
code_2_extract(code6)
code_2_extract(code7)
code_2_extract(code8)
code_2_extract(code9)
code_2_extract(code10)
code_2_extract(code11)
code_2_extract(code12)

```

**Comment by chlebowa** [comment]
It does indeed :+1: 

Proper review will have to wait, though.

Merry Christmas :christmas_tree:

File: R/utils-get_code_dependency.R [hunk 1]
```diff
@@ -35,17 +35,22 @@ get_code_dependency <- function(code, names, check_names = TRUE) {
...
 
   if (check_names) {
     # Detect if names are actually in code.
-    symbols <- unique(pd[pd$token == "SYMBOL", "text"])
-    if (!all(names %in% symbols)) {
+    symbols <- unlist(lapply(calls_pd, function(call) call[call$token == "SYMBOL", "text"]))
+    if (any(pd$text == "assign")) {
+      assign_calls <- Filter(function(call) any(call$token == "SYMBOL_FUNCTION_CALL" & call$text == "assign"), calls_pd)
+      ass_str <- unlist(lapply(assign_calls, function(call) call[call$token == "STR_CONST", "text"]))
+      ass_str <- gsub("'", "", ass_str)
```
Comment (chlebowa) [inline R/utils-get_code_dependency.R:46]: What about regular (escaped) quotation marks? How about `gsub("['\"]", "", ass_str)` or `gsub("^['\"]|['\"]$", "", ass_str)`?

File: R/utils-get_code_dependency.R [hunk 2]
```diff
@@ -35,17 +35,22 @@ get_code_dependency <- function(code, names, check_names = TRUE) {
...
   pd <- utils::getParseData(code)
+  calls_pd <- extract_calls(pd)
 
   if (check_names) {
     # Detect if names are actually in code.
-    symbols <- unique(pd[pd$token == "SYMBOL", "text"])
-    if (!all(names %in% symbols)) {
+    symbols <- unlist(lapply(calls_pd, function(call) call[call$token == "SYMBOL", "text"]))
+    if (any(pd$text == "assign")) {
+      assign_calls <- Filter(function(call) any(call$token == "SYMBOL_FUNCTION_CALL" & call$text == "assign"), calls_pd)
```
Comment (chlebowa) [inline R/utils-get_code_dependency.R:44]: Consider creating a function like `is_assign` to use here and then again in `extract_occurence`.

File: R/utils-get_code_dependency.R [hunk 3]
```diff
@@ -167,8 +172,19 @@ extract_occurrence <- function(calls_pd) {
...
-        call_pd_lim <- call_pd[-c(1:which(assign_call)), ]
-        sym <- call_pd_lim[call_pd_lim$token == "STR_CONST", "text"]
+        # Check if parameters were named.
+        if (any(call_pd$token == "SYMBOL_SUB")) {
+          params <- call_pd[call_pd$token == "SYMBOL_SUB", "text"]
+          if ("x" %in% params) {
+            pos <- which(params == "x")
+          } else {
+            pos <- length(params) + 1
+          }
```
Comment (chlebowa) [inline R/utils-get_code_dependency.R:182]:
```suggestion
          pos <- match("x", params, nomatch = length(params) + 1L)
```

File: tests/testthat/test-get_code.R [hunk 4]
```diff
@@ -136,6 +136,49 @@ testthat::test_that("get_code with datanames is possible to output the code for
...
+  )
+  tdata <- eval_code(teal_data(), code)
+  datanames(tdata) <- c("a", "b")
+  testthat::expect_identical(
+    get_code(tdata, datanames = "b"),
+    paste("b <- 2", "eval(expression({\n    b <- b + 2\n}))", sep = "\n")
+  )
+})
+
+testthat::test_that("get_code returns result of length for non-empty input", {
```
Comment (chlebowa) [inline tests/testthat/test-get_code.R:167]:
```suggestion
testthat::test_that("get_code returns result of length 1 for non-empty input", {
```

File: tests/testthat/test-get_code.R [hunk 5]
```diff
@@ -195,63 +238,17 @@ testthat::test_that("@linksto tag indicate affected object if object is assigned
...
-  }
-)
-
-testthat::test_that("get_code with datanames can extract the code when using <<-", {
-  code <- c(
-    "a <- 1",
-    "b <- a",
-    "b <<- b + 2"
-  )
+testthat::test_that("get_code extracts a line with assign where object is passed as character", {
```
Comment (chlebowa) [inline tests/testthat/test-get_code.R:241]: Please refine this description. "object is passed as character" is not readily understandable.

File: tests/testthat/test-get_code.R [hunk 6]
```diff
@@ -653,3 +650,19 @@ testthat::test_that("get_call data call is returned when data name is provided a
     )
   )
 })
+
+testthat::test_that(
+  "get_code with datanames can extract the code when function creates an object which is used only on rhs",
```
Comment (chlebowa) [inline tests/testthat/test-get_code.R:655]: Please refine this description, it does not convey what case is tested.

**Review by chlebowa — CHANGES_REQUESTED** [review]
`assign('b', 5)` works fine but `assign(\"b\", 5)` fails in unit test `"get_code with datanames can extract the code for assign function"`.

File: R/utils-get_code_dependency.R (hunk 1 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R:46]: will double check, thanks :)!

**Comment by m7pr** [comment]
Thanks for review @chlebowa ! Will use your feedback to improve!

File: R/utils-get_code_dependency.R (hunk 1 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R]: Just pushed a fix for that, and also extended the mentioned test ff6c2fd37ef302e56a42a7950c3dad7b1d4f909d

File: tests/testthat/test-get_code.R (hunk 5 above)
Comment (m7pr) [reply tests/testthat/test-get_code.R]: this was a redundant test that was already covered in other test so I decided to remove f9e05c015dc35525702e17191feaa1ee4773b280

File: tests/testthat/test-get_code.R (hunk 6 above)
Comment (m7pr) [reply tests/testthat/test-get_code.R]: 09b02dda70eb153ffa97f917956001d381451f17

File: R/utils-get_code_dependency.R (hunk 2 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R]: Great improvement. I created `is_symbol` function and also used that in situation where we seek for `data` function as well b56c5c426c96dac78130a395a7014e16b2b4d369

File: R/utils-get_code_dependency.R [hunk 7]
```diff
@@ -54,6 +59,10 @@ get_code_dependency <- function(code, names, check_names = TRUE) {
   as.character(code[unique(c(lib_ind, ind))])
 }
 
+is_symbol <- function(call_pd, symbol) {
```
Comment (chlebowa) [inline R/utils-get_code_dependency.R:62]: Please add a comment explaining what exactly this function does. I hardly remember what I envisioned two weeks ago myself.

Also, `@internal`.

File: R/utils-get_code_dependency.R [hunk 8]
```diff
@@ -163,17 +172,24 @@ extract_occurrence <- function(calls_pd) {
     calls_pd,
     function(call_pd) {
       # Handle data(object)/data("object")/data(object, envir = ) independently.
-      data_call <- call_pd$token == "SYMBOL_FUNCTION_CALL" & call_pd$text == "data"
+      data_call <- is_symbol(call_pd, "data")
       if (any(data_call)) {
```
Comment (chlebowa) [inline R/utils-get_code_dependency.R:176]: I think it's safe to do 
```suggestion
      if (any(is_symbol(call_pd, "data"))) {
```

By the way, it is vital that `is_symbol` always return a logical vector of length > 0, never `NULL`, `logical(0)`, `NA`. Please make sure it always happens in the context the function is used or introduce some argument checks to lock it in.

**Review by chlebowa — CHANGES_REQUESTED** [review]

File: R/utils-get_code_dependency.R (hunk 8 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R:176]: `data_call` is used further after the `if` statement so I will keep it as a variable, and not a check in the `if`.

```r
      data_call <- is_symbol(call_pd, "data")
      if (any(data_call)) {
        sym <- call_pd[which(data_call) + 1, "text"]
```

Comment (m7pr) [reply R/utils-get_code_dependency.R:176]: > By the way, it is vital that is_symbol always return a logical vector of length > 0, never NULL, logical(0), NA. Please make sure it always happens in the context the function is used or introduce some argument checks to lock it in.

I don't think there is no need to check that. `is_symbol` returns vector of logicals of length 0 or greater. Then it is used in `any` that returns `FALSE` on `logical(0)` which is great, and `TRUE` on `logical(n)` if any element is equal to `TRUE`. I don't think we need any special treatment of edge cases as it's vectorized logical operation on two vectors

`call_pd$token == "SYMBOL_FUNCTION_CALL" & call_pd$text == symbol`

and those vectors can either be empty or of type `character` (and never seen it to be `NA`).

This was already tested in the battlefield in `tests/` folder anyway : )

Just to show what are the results of usage please see below, but I think testing this is an overkill. I didn't put anything extra in the function besides `checkmate::check_string(symbol)`

```r
call_pd_1 <- data.frame(
 token = character(0),
 text = character(0)
)

call_pd_2 <- data.frame(
 token = 'SYMBOL_FUNCTION_CALL',
 text = 'data1'
)

call_pd_3 <- data.frame(
 token = c('SYMBOL', 'SYMBOL_FUNCTION_CALL'),
 text = c('iris', 'data')
)

is_symbol(call_pd_1, 'data')
> logical(0)
is_symbol(call_pd_2, 'data')
> [1] FALSE
is_symbol(call_pd_3, 'data')
> [1] FALSE  TRUE
any(is_symbol(call_pd_1, 'data'))
> [1] FALSE
any(is_symbol(call_pd_2, 'data'))
> [1] FALSE
any(is_symbol(call_pd_3, 'data'))
> [1] TRUE
any(NULL)
> [1] FALSE
any(logical(0))
> [1] FALSE

```

File: R/utils-get_code_dependency.R (hunk 7 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R:72]: Hey @chlebowa thanks for taking the time before New Years and now to review that!
I added a documentation to `is_symbol` eca80e690dfd3bdc92b08b953aec9024fd3adc9b

There is a lot of text added as a pay-off of the substitution of 3 lines with a new function, but in few months it's gonna be useful to understand what's the purpose of the function, even though it's a one liner function.

File: R/utils-get_code_dependency.R (hunk 8 above)
Comment (m7pr) [reply R/utils-get_code_dependency.R:189]: Ok, I ended up with such checks for `is_symbol` functions

```r
checkmate::check_string(symbol)
checkmate::check_data_frame(call_pd)
checkmate::check_names(call_pd, must.include = c('token', 'text'))
 ```
 
 I think it's enough for this simple function.

Comment (chlebowa) [reply R/utils-get_code_dependency.R:189]: Fair enough, 
```
> any(NULL)
[1] FALSE
> any(logical(0))
[1] FALSE
```

Comment (chlebowa) [reply R/utils-get_code_dependency.R:189]: > `data_call` is used further

You're right, I missed that.

**Review by chlebowa — APPROVED** [review]
:+1:

**Comment by m7pr** [comment]
Hey @chlebowa one last idea, what if we change the name from `is_symbol` to `is_function` ? As we actually use it with `SYMBOL_FUNCTION_CALL` token statement

**Pull request merged**